DynamicActiveQuery.php

```diff
@@ -27,6 +28,45 @@ class DynamicActiveQuery extends ActiveQuery
      */
     private $dynamicColumn;
 
+    private $wrapped = false;
+
+    /**
+     * Convert index value to closure, that will get decoded dynamic attribute, in case if indexing attribute is dynamic
+     * @param callable|string $column
+     * @return $this
+     */
+    public function indexBy($column)
+    {
+        if (!$this->asArray) {
+            return parent::indexBy($column);
+        }
+
+        /** @var \yii\db\ActiveRecord $modelClass */
+        $modelClass = $this->modelClass;
+        $this->indexBy = function ($row) use ($column, $modelClass) {
+            if (isset($row[$column])) {
+                return $row[$column];
+            }
+
+            if (!method_exists($modelClass, 'dynamicColumn')) {
+                throw new Exception("Indexable column {$column} does not exist");
+            }
+
+            $dynamicColumn = $modelClass::dynamicColumn();
+            if (!isset($row[$dynamicColumn])) {
+                throw new Exception("Dynamic column {$dynamicColumn} does not exist");
+            }
+
+            $dynamicAttributes = DynamicActiveRecord::dynColDecode($row[$dynamicColumn]);
+            $value = $this->getDotNotatedValue($dynamicAttributes, $column);
+
+            // in case if dynamic column does not exist for this row - use PK
```

Comment:
ok, yeah, i wasn't sure too, but you should decide whether indexing should be supported